First pass at primitive tool: Cylinders only.

jitspoe opened the pull request:
At the moment, it only creates cylinders.  There are 2 tools.  One in the toolbar like normal tools, and one that can be used to convert an existing brush to a primitive (BSP editor style).  Also added the ability to delete brushes from within a tool (only used in primitive tool now, but could be useful in other cases).

Related feature request: https://github.com/TrenchBroom/TrenchBroom/issues/398

kduske commented:
I haven't tried this yet, but here's some initial feedback based on your description. I would not have added a new tool for this. Generally, I try to avoid separate tools or modes in TB as much as possible, as switching tools and modes impedes the flow state by interrupting their workflow.

For primitive creation, I would have tested out extending the current brush creation tool where you draw a brush by defining its bounding box with the mouse.

Extend the toolbar above the 3D/2D editing views (which now only contains the "move objects by" controls) with the following controls:
- a selector for a basic shape (cuboid, cylinder, ...)
- a way to select a main orientation axis (e.g. + X axis, - X axis, ...)
- optionally, a way to set a rotation about the main orientation axis

Now, when you draw a brush by dragging with the left mouse, instead of just creating a cuboid, you create whatever shape the user chose with the parameters they set. They get a real time preview of the shape and it works in the same way as the tool they already know (you can draw on top of other objects etc.).

One thing to change, maybe in a separate PR, is to draw the preview fully textured instead of just gray like now.

If the small toolbar gets too cluttered, I would consider putting the "move objects by" function elsewhere, like in a menu with a popup dialog or something.

What do you think about this?

jitspoe commented:
I'm not sure what the best user experience would be to be honest.  I posted some gifs in https://github.com/TrenchBroom/TrenchBroom/issues/398 to show the current behaviors.  I think having a separate tool is useful for the sake of visibility and intuitiveness.  Currently it more or less behaves the same as no tool selected but with a bunch of primitive options.  It could be effectively a sub-tool of the default tool.  I would like to add a lot of other functionality, too, so it might be a bit overwhelming to have that in the default tool.

Initially, I started implementing it more like BSP had it, with a separate dialogue that pops up where you convert an existing brush to another shape, but some people on discord suggested I make it a tool in the toolbar that you could then click and drag.  I still have that in the PR as well, but it would probably make more sense to consolidate that functionality into one location.

kduske commented:
To me it looks like one of the versions is already close to what I proposed: https://user-images.githubusercontent.com/7751182/249394406-c03c68fb-b16b-4226-bba2-5821581d7bec.gif

Is that what you submitted in this PR?

jitspoe commented:
> To me it looks like one of the versions is already close to what I proposed: https://user-images.githubusercontent.com/7751182/249394406-c03c68fb-b16b-4226-bba2-5821581d7bec.gif
> 
> Is that what you submitted in this PR?

I actually have both versions in the PR.  Also, just updated the PR so it compiles with the latest TB master.

kduske commented:
I think while both workflows have their merit, we only need one. The one where the user draws the primitive directly into the views fits better with TB's general workflow IMO. It will be a challenge to keep the whole thing simple and quick, though. I think we should avoid too many options and settings.

jitspoe commented:
So after actually using it for a bit, here are my thoughts:
Sometimes it's difficult to get the correct size in one go with click and drag, especially if you're trying to do a really big curve.  I think it might make sense to have controls kind of like the scaling tool for post-creation adjustment, and when doing those adjustments, it would re-calculate the values so they can be snapped accurately.  For example, if you created a cylinder too small, had it snapped to grid, then scaled it up slightly, it would no longer be snapped, and if you snapped it, it might look off, since it was created snapped to a grid when a different size.

It might also be good to have exact values that can be input with an apply button (kind of like how you can set the rotation position and whatnot) to replace the selected brush with the new values.  That would get us the functionality of both workflows but in one location.  Could potentially have some of this stuff tucked behind an "advanced" UI if it gets to be too cluttered.

kduske commented:
I can understand that there are tradeoffs. But rather than offer two different ways of doing the same, I would like to offer only one tool. Maybe we could offer a command that lets the user easily adjust a brush that they created by left dragging. Essentially, you would create your cylinder by drawing it, and if it turned out wrong, hit a hotkey and be able to readjust it.

What do you think about that? That could be totally independent from this PR, too.

bstordeur commented:
I just tested the PR a bit and read the discussion above, cool stuff!
Some random thoughts : 

I agree with Kristian on the choice of direction overall, the extra window with parameters feels like something from the 90s, and the kind of thing new modeling apps are gradually removing. It's really clunky and breaks up flow, but it also gives you the most control, so it's quite tricky to replace completely.

My first thought is, why is it called "Primitive Brush Tool"? As of now it's very much limited to making cylinders.
I guess we could imagine this tool being expanded to other shapes, but that seems tricky with the UI. If that's what is intended, I could imagine there being a drop-down list on the left (next to "number of sides") that changes the parameters available.
I'd have some ideas of useful primitives if this is the direction you go in (like pie-cylinder, flat torus with thickness params, etc.)

Otherwise I feel the implementation feels pretty good, I'm not sure adding something to edit the result is really necessary, as most of the time just deleting the brush and starting over will go just as fast? Instead it could maybe be like the brush creation tool where you have to validate the result by pressing enter, which means until then you can change the # of sides etc. and it'll update.
Could also imagine some context hotkeys, for example in Blender for some operations, you'll be able to use the mouse-wheel while dragging out a bevel to add/remove edges. Maybe something similar could be done here.

I think a checkbox could be added to force proportional shapes (so the cylinder stays a circle), as that's what users will want most of the time, and depending on your grid size it can be awkward.

One thing that would be really powerful, is to apply the active texture in a "smart" way.
With a bit of maths, calculate the circumference of the cylinder, find the closest ratio the current texture should be so that it loops around the cylinder n times while staying as close to the base texel ratio as possible. This could also be a optional checkbox.

I'm a bit biased since I map only for Quake, and cylinders are a tricky subject because we're often required to use weird approximations (the "czg" curve) to build complex curved structures, maybe a particular primitive type could be created for this.
But overall I think what you have in the PR with the Primitive Brush Tool is already really close to be a worthy addition to TB! New users are always asking how to make quick pillars or pipes, and this tool would really help.

jitspoe commented:
Yeah, the plan is to add support for multiple primitives (cones, tubes, spheres, archs, etc).  Probably with a drop-down as you suggested.  I come from the BSP world of mapping, and it has a nice dropdown to make a bunch of different things:
![image](https://github.com/TrenchBroom/TrenchBroom/assets/7751182/4a9c964b-d306-4133-8c84-e352dc1fcf52)

I just wanted to start off with 1 to begin with before doing a bunch of UI work and having to redo it for a different workflow, as it's not super trivial.  Figured I'd get the cylinders in then add more stuff later once we were happy with the workflow.

Smart UV mapping is definitely something I'd like to have as well.  The reason I was suggesting some post-editing options is because it can be challenging to work with large-scale stuff.  For example, if you're making a giant arch or circular room, you might want something that's exactly a 512 unit radius, but it needs to line up with something on a 16 unit grid (just something I personally ran into), and clicking and dragging isn't super viable.  In any case, that kind of stuff can be added later.

Hopefully I can find time to put together a new PR.

jitspoe commented:
Closing this.  Made a completely new pull request that's a bit cleaner and addresses some of the issues with this one: https://github.com/TrenchBroom/TrenchBroom/pull/4350
